tools/check_diff_coverage.py

```
@@ -0,0 +1,173 @@
+
+
+def changed_lines(base_sha: str, head_sha: str, source_root: str) -> dict[str, set[int]]:
+    """Return added or modified line numbers grouped by production file.
+
+    Git is invoked with an argument vector and ``shell=False``. File names from
+    the diff are parsed as text only; they are never interpolated into a shell.
+    """
+
+    base = _validated_git_object(base_sha, argument="base_sha")
+    head = _validated_git_object(head_sha, argument="head_sha")
+    completed = subprocess.run(
```

Ruff (0.16.1)

[error] `subprocess` call: check for execution of untrusted input

(S603)